Added Prompt System to Editor, Fixed Bug with Text Underneath Token

Alko-li opened the pull request:
Added a prompt system to the program editor to notify user of invalid actions, and also fixed bug where very long underlying text in a token slot appeared even when token is placed over slot.

Alko-li committed:
Merge pull request #1 from BrandonShockley/master

Merge changes from BrandonShockley

Alko-li committed:
Merge pull request #2 from BrandonShockley/master

Merge pull request from BrandonShockley

Commit:
Started on Prompt System, fixed bugs

Fixed bugs with inserting tokens into text fields with long text

Commit:
Merge branch 'master' of https://github.com/Alko-li/Decontainment

Commit:
Added Prompt System to Editor

Added a prompt system to the program editor to warn for invalid actions.

Commit:
Changed number of ticks prompt stays up

Commit:
Revert "Merge branch 'master' of https://github.com/Alko-li/Decontainment"

This reverts commit 69bfb37eeeab18ea36d878ef2e12757324144d53, reversing
changes made to 8a541062f4171eebd8bebf5cc3fe56bf07a976cf.

Commit:
Revert "Started on Prompt System, fixed bugs"

This reverts commit 8a541062f4171eebd8bebf5cc3fe56bf07a976cf.

Commit:
Merge branch 'master' into prompt

Commit:
Revert "Merge branch 'master' into prompt"

This reverts commit 3f8f143d6426677ea42f6070ad34e7d15a3b7684, reversing
changes made to 9c04cb985a0821086fee7660841b861c3ca06bbf.

Commit:
Revert "Revert "Started on Prompt System, fixed bugs""

This reverts commit c9db3768dbc8e8be97046c3366f9054491dda56e.

Commit:
Removed extra PromptSystem.cs

BrandonShockley commented on Assets/Decontainment/BotData/Programs/Dodge.txt:
```diff
@@ -1,28 +0,0 @@
-Inf: 1000000000
```
Don't delete this file

BrandonShockley commented on Assets/Decontainment/Scenes/Editor.unity:
```diff
@@ -1379,6 +1380,91 @@ RectTransform:
   m_PrefabInstance: {fileID: 5389810365148441542}
   m_PrefabAsset: {fileID: 0}
+--- !u!1 &425484728846292843
+GameObject:
```
This Banner game object isn't linked to the Banner prefab you made

BrandonShockley commented on Assets/Decontainment/Scenes/Test.unity:
```diff
@@ -1278,6 +1279,21 @@ PrefabInstance:
       propertyPath: m_textInfo.pageCount
       value: 0
       objectReference: {fileID: 0}
+    - target: {fileID: 3111098471529109542, guid: cbac44966c4dc4f4a8f3885057157673,
```
This file should not be modified.

BrandonShockley commented on Assets/Decontainment/UI/CodeEditor/Prefabs/Banner.prefab:
```diff
@@ -0,0 +1,276 @@
+%YAML 1.1
+%TAG !u! tag:unity3d.com,2011:
+--- !u!1 &2249473649416014478
+GameObject:
```
Move this prefab to UI/Misc/Prefabs

BrandonShockley commented on Assets/Decontainment/UI/CodeEditor/Prefabs/Banner.prefab:
```diff
@@ -0,0 +1,276 @@
+  m_StaticEditorFlags: 0
+  m_IsActive: 1
+--- !u!224 &1585251172312195815
+RectTransform:
```
Do not scale UI components. Set the scales all back to 1 and adjust the rest of the sizing fields to get what you want.

It looks like you were going for adding padding to all sides of the banner, which is a good idea. For that, I suggest resetting the text component's margins back to 0 (check the extra settings dropdown) then setting the left and right, and top and bottom fields of the rect transform to equal values until you get what you want. Something like this:
![Capture](https://user-images.githubusercontent.com/9877377/74607221-0318bc00-50a5-11ea-8f4f-c8931f9eeea5.PNG)

BrandonShockley commented on Assets/Decontainment/UI/CodeEditor/Prefabs/Banner.prefab:
```diff
@@ -0,0 +1,276 @@
+  m_StaticEditorFlags: 0
+  m_IsActive: 1
+--- !u!224 &7778551349502851045
+RectTransform:
```
The Banner's rect transform should use anchors to span itself across the screen instead of a static size. Set the max X anchor to 1 to accomplish this, then set the "right" field to 0. Additionally, set the pivot Y and pos Y to 0 to ensure the banner is always at the bottom of the screen.

BrandonShockley commented on Assets/Decontainment/UI/CodeEditor/Scripts/Banner.cs:
```diff
@@ -0,0 +1,46 @@
+public class Banner : MonoBehaviour
+{
+	[SerializeField]
+	private int ticks = 5;
```
Your timing should not be based on frames. Update is called every frame, so your counter will be linked to the framerate, which can vary. You should modify this system to work with seconds. Change `ticks` to `duration` and change it and the counter to a float value. Every update, you can subtract `Time.deltaTime` from counter and check if it's below zero. The rest of your logic should be fine.

BrandonShockley commented on Assets/Decontainment/UI/CodeEditor/Scripts/Banner.cs:
```diff
@@ -0,0 +1,46 @@
+    private Image image;
+    private TextMeshProUGUI text;
+
+    protected Color defaultColor = new Color(0, 0, 0, 0);
```
It doesn't look like this is used. If it isn't, remove it.

BrandonShockley commented on Assets/Decontainment/UI/CodeEditor/Scripts/Banner.cs:
```diff
@@ -0,0 +1,46 @@
+    void Update()
+    {
+		if (counter >= 0)
+        {
```
This bracket should be on the same line as the if statement it corresponds to. This is mentioned in the style guide.

BrandonShockley commented on Assets/Decontainment/UI/CodeEditor/Scripts/Banner.cs:
```diff
@@ -0,0 +1,46 @@
+            text.enabled = true;
+            counter = counter - 1;
+        } else
+        {
```
Bracket should be on same line as else

BrandonShockley commented on Assets/Decontainment/UI/CodeEditor/Scripts/Banner.cs:
```diff
@@ -0,0 +1,46 @@
+    protected Color defaultColor = new Color(0, 0, 0, 0);
+
+    // Start is called before the first frame update
+    void Start()
```
This initialization should probably happen in the Awake() method

BrandonShockley commented on Assets/Decontainment/Singletons/Scripts/PromptSystem.cs:
```diff
@@ -0,0 +1,63 @@
+using TMPro;
+using UnityEngine.UI;
+
+public class PromptSystem : PersistentSingleton<PromptSystem>
```
`PersistentSingleton`s are game objects that are not destroyed when a new scene is loaded. The PromptSystem currently acquires a reference to a banner object in `Awake`. When a new scene is loaded, the banner object will be destroyed, but the PromptSystem will persist. If one then tries to call `PromptSystem.Instance.PromptInvalidAction()`, it will call `banner.Activate()`, but banner will be null and the code will break. Therefore, this solution does not work.

I suggest having PromptSystem inherit from `SceneSingleton` then creating a game object with the PromptSystem component in the editor scene. `SceneSingleton`s are destroyed and recreated when a new scene is loaded, so this would fix your issue.

BrandonShockley commented on Assets/Decontainment/Singletons/Scripts/PromptSystem.cs:
```diff
@@ -0,0 +1,63 @@
+    void Awake()
+    {
+        banner = FindObjectOfType<Banner>();
+        obj = banner.gameObject;
```
You don't need obj here. Just use banner.GetComponent when getting component references.

Commit:
Fixed issues with PromptSystem

Added back Dodge.txt, made PromptSystem a SceneSingleton, made Banner update using Time.deltaTIme rather than ticks

BrandonShockley commented on Assets/Decontainment/UI/CodeEditor/Scripts/Banner.cs:
```diff
@@ -0,0 +1,45 @@
+    {
+        if (counter >= 0)
+        {
+            counter -= Time.fixedDeltaTime;
```
This should use Time.deltaTime, not Time.fixedDeltaTime

BrandonShockley reviewed (changes requested):
Some of my previous comments still need to be addressed. Also you need to fix the merge conflicts. I can try to help with this at the meeting if you'd like.

Commit:
Update Banner.cs

Commit:
Merge remote-tracking branch 'upstream/master' into prompt

Commit:
Revert "Merge remote-tracking branch 'upstream/master' into prompt"

This reverts commit 6a4082d5e2121dc86d289f5fa180b4b8e96d60b4, reversing
changes made to c05ceabc4850cc190ad1a46d0ccba3b67a9b8f8e.

Commit:
Merge remote-tracking branch 'upstream/master' into prompt

Commit:
Test Branch

Commit:
Added prompt system

Commit:
Added Test Changes to Prompt Branch

Commit:
Merge remote-tracking branch 'upstream/master' into prompt

BrandonShockley commented:
I'm going to go ahead and merge this. It's been a few weeks of unproductive back and forth change requests. There's still style things you should have addressed, but I can fix those up myself. Everything functional is working.
